[10.0] [FIX] point_of_sale: allow to create more than one rescue session

mart-e opened the pull request:
Fixes #15101 (cf comment https://github.com/odoo/odoo/issues/15101#issuecomment-273208287 for details)

The 'RESCUE FOR' is no longer translated as it was not possible to detect it when translated
> provided we are careful not to translate the keyword.

With the first commit one rescue session is create per order

The second use make sure to reuse an already open rescue session (but can be skipped if want to avoid issue raised at e892c574825c19598fe3c243fcb651f8b6d50d0f).

#15864 for master

odony commented on addons/point_of_sale/models/pos_session.py:
```diff
@@ -126,12 +126,20 @@ def _compute_cash_all(self):
+        if self.search_count([
+                ('state', '!=', 'closed'),
+                ('config_id', '=', self.config_id.id),
+                ('name', 'not like', '%RESCUE FOR%'),
```
I think the `%` signs are redundant with the default behavior of the `not like` operator, even if they won't break anything ;-)
Same thing in `_check_unicity`

mart-e replied:
oh right, only for matching in the middle

jorenvo commented on addons/point_of_sale/models/pos_session.py:
```diff
@@ -126,12 +126,20 @@ def _compute_cash_all(self):
+        if self.search_count([
+                ('state', 'not in', ('closed', 'closing_control')),
+                ('user_id', '=', self.user_id.id),
+                ('name', 'not like', 'RESCUE FOR'),
```
Maybe for consistency we could turn this into:

```python
'!', ('name', '=like', '%(RESCUE FOR%)'),
```

as well.

odony replied:
On second thought, I'd rather not use `=like` at all anywhere, it makes the domains harder to read. And if someone really decides to use "RESCUE FOR" as the prefix in their POS sequence, I'd say they asked for trouble and are likely to get some anyhow.

jorenvo commented on addons/point_of_sale/models/pos_session.py:
```diff
@@ -126,12 +126,20 @@ def _compute_cash_all(self):
+        if self.search_count([
+                ('state', '!=', 'closed'),
+                ('config_id', '=', self.config_id.id),
+                ('name', 'not like', 'RESCUE FOR'),
```
Same as above.

odony replied:
Thanks @jorenvo for testing the PR properly! :-)

jorenvo commented on addons/point_of_sale/models/pos_session.py:
```diff
@@ -126,12 +126,20 @@ def _compute_cash_all(self):
     @api.constrains('user_id', 'state')
     def _check_unicity(self):
         # open if there is no session in 'opening_control', 'opened', 'closing_control' for one user
-        if self.search_count([('state', 'not in', ('closed', 'closing_control')), ('user_id', '=', self.user_id.id)]) > 1:
```
Here we change the unicity constraints to allow > 1 session to be open which is what we want but this requires two more changes to avoid tracebacks on the POS dashboard:

```
Traceback (most recent call last):
...
  File "/home/odoo/odoo/addons/point_of_sale/models/pos_config.py", line 141, in _compute_current_session
    pos_config.current_session_id = session
  File "/home/odoo/odoo/odoo/fields.py", line 887, in __set__
    value = self.convert_to_cache(value, record)
  File "/home/odoo/odoo/odoo/fields.py", line 1902, in convert_to_cache
    raise ValueError("Wrong value for %s: %r" % (self, value))
ValueError: Wrong value for pos.config.current_session_id: pos.session(23, 22)
```

```
Traceback (most recent call last):
...
  File "/home/odoo/odoo/addons/point_of_sale/models/pos_config.py", line 163, in _compute_current_session_user
    pos_config.pos_session_username = pos_config.session_ids.filtered(lambda s: s.state == 'opened').user_id.name
  File "/home/odoo/odoo/odoo/fields.py", line 863, in __get__
    record.ensure_one()
  File "/home/odoo/odoo/odoo/models.py", line 4803, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
ValueError: Expected singleton: pos.session(23, 22)
```

On ```pos.config``` both ```_compute_current_session()``` and ```_compute_current_session_user()``` need to only search on non-rescue sessions. So for ```_compute_current_session()``` something like:

```python
and '(RESCUE FOR ' not in r.name
```

can be added to the ```filtered``` lambda. Similar for ```_compute_current_session_user()```.

This is analogous to how it was in 9.0: [_compute_current_session()](https://github.com/odoo/odoo/blob/9.0/addons/point_of_sale/point_of_sale.py#L48), [_compute_current_session_user()](https://github.com/odoo/odoo/blob/9.0/addons/point_of_sale/point_of_sale.py#L80)

mart-e replied:
Ok to match the v9 version

odony commented on addons/point_of_sale/models/pos_order.py:
```diff
@@ -66,10 +66,18 @@ def _get_valid_session(self, order):
                         order['name'],
                         order['amount_total'])
+        rescue_session = PosSession.search([
+            ('name', '=like', '%(RESCUE FOR %(session)s)' % {'session': closed_session.name}),
```
I reviewed the patch before that last change, and I think the previous version was fine with `'like', '(RESCUE FOR %(session)s)' % {'session': closed_session.name})`. The use of `=like` was an attempt to be *even more restrictive*, but there is virtually no risk of collision here. It will also not trigger the bug with the Python format string.

I would like to keep the name of the session in the criterion, because it serves as an implicit selector for the correct POS config_id (which the master PR has). Alternatively, if we only search for "(RESCUE FOR" we need to add a criterion on the config_id.

jorenvo replied:
Ah right, I missed that in master the 'config_id' was added in the search. Right well, it doesn't matter much I suppose. If we do it this way there will be a difference in behavior in 10.0 and master when there are multiple rescue sessions open for the same pos.config.

I prefer to take master as an example of how it should be (with the ```rescue``` field):

```python
('state', 'not in', ('closed', 'closing_control')),
('rescue', '=', True),
('config_id', '=', closed_session.config_id.id),
```

which for 10.0 would become something like:

```python
('state', 'not in', ('closed', 'closing_control')),
('name', 'like', '(RESCUE FOR '), # fake rescue field
('config_id', '=', closed_session.config_id.id),
```

I'm not sure what the advantage is of adding additional divergence between both versions but anyway, you guys decide. I don't think there is a major downside to both approaches.

odony commented on addons/point_of_sale/models/pos_order.py:
```diff
@@ -66,10 +66,18 @@ def _get_valid_session(self, order):
+            ('state', 'not in', ('closed', 'closing_control')),
+        ], limit=1)
+        if rescue_session:
+            _logger.warning('reusing recovery session %s for saving order %s', (rescue_session.name, order['name']))
```
Good point, I missed that. But please do not interpolate yourself, let logging do it for you, as you intended.

In fact this should have made the tests fail, perhaps we should extend the coverage to the rescue flow...

odony commented on addons/point_of_sale/models/pos_order.py:
```diff
@@ -66,10 +66,18 @@ def _get_valid_session(self, order):
             'config_id': closed_session.config_id.id,
-            'name': _('(RESCUE FOR %(session)s)') % {'session': closed_session.name},
+            'name': '(RESCUE FOR %(session)s)' % {'session': closed_session.name},
             'rescue': True,    # avoid conflict with live sessions
```
Sure, and it would make the build orange as soon as we extend the coverage ;-)

mart-e commented:
Thanks for the review !
I have added longer tests, creating multiple orders through the `create_from_ui` method (previous ones were only calling `create` directly), even after closing the session.

Note that I had to add the rescue check also on the `/pos/web` controller to prevent opening a rescue session (and break on the `ensure_one` of `login` if there is opened rescue sessions).

My first idea was to prevent opening a new session if rescue sessions are still open but with FTK (the one we implemented the rescue in the first place), they used rescue flag as a way to keep selling in a new one when they could not close one (e.g. rounding issue).
So in the end, I just ignored rescue sessions from the login.

jorenvo commented on addons/point_of_sale/tests/test_point_of_sale_flow.py:
```diff
@@ -411,3 +411,138 @@ def test_order_to_invoice(self):
+           'statement_ids': [[0,
+             0,
+             {'account_id': self.env.user.partner_id.property_account_receivable_id.id,
+              'amount': 1.3800000000000001,
```
And the payment has the same rounding weirdness, which if it really is a bug in ```round_precision()``` makes sense I guess.

jorenvo commented:
I like the extra tests you added :).

> So in the end, I just ignored rescue sessions from the login.

Sounds good to me. It would be weird for a user to continue using a rescue session.

For me this is good to merge.

jorenvo commented:
Oh, just saw that the test doesn't pass yet. I guess you're still working on it. It's cause the backend recomputes ```amount_total``` on ```pos.order```. So you have to make sure that the different ```lines``` in your test data match what you want the total price to be.

mart-e commented on addons/point_of_sale/tests/test_point_of_sale_flow.py:
```diff
@@ -411,3 +411,138 @@ def test_order_to_invoice(self):
+          'to_invoice': False}
+
+        zucchini_order = {'data':
+          {'amount_paid': 1.3800000000000001,
```
exactly, got it from the pos itself, we are closer to the real pos data

mart-e commented:
Strange, the test pass on mu machine but not on the runbot :-/

mart-e commented:
Ok the test was failing due to different in the local taxes and the one on the runbot ;-)
Thanks for the review !
Will update #15864 once it is forward ported
